synapse/push/__init__.py

```diff
@@ -266,23 +281,64 @@ def dispatch_push(self, p, tweaks):
         """
         pass
 
-    def reset_badge_count(self):
-        pass
+    @defer.inlineCallbacks
+    def update_badge(self):
+        new_badge = yield self._get_badge_count()
+        if self.badge != new_badge:
+            self.badge = new_badge
+            yield self.send_badge(self.badge)
 
-    def presence_changed(self, state):
+    def send_badge(self, badge):
         """
-        We clear badge counts whenever a user's last_active time is bumped
-        This is by no means perfect but I think it's the best we can do
-        without read receipts.
+        Overridden by implementing classes to send an updated badge count
         """
-        if 'last_active' in state.state:
-            last_active = state.state['last_active']
-            if last_active > self.last_last_active_time:
-                self.last_last_active_time = last_active
-                if self.has_unread:
-                    logger.info("Resetting badge count for %s", self.user_id)
-                    self.reset_badge_count()
-                    self.has_unread = False
+        pass
+
+    @defer.inlineCallbacks
+    def _get_badge_count(self):
+        membership_list = (Membership.INVITE, Membership.JOIN)
+
+        room_list = yield self.store.get_rooms_for_user_where_membership_is(
+            user_id=self.user_id,
+            membership_list=membership_list
```

Comment:
Any reason not to simply `membership_list=(Membership.INVITE, Membership.JOIN),`?